### PR TITLE
Add examples for using SMDDP in sharded data parallel training with FSDP and Deepspeed 

### PR DESCRIPTION
**Pull request by ruhanprasad**
*Issue #, if available:*

*Description of changes:*
SageMaker Distributed Data Parallel (SMDDP) library is coming out with a new release starting in PyTorch 2.0.1 DLC that will allow customers to accelerate sharded data parallel training jobs with PyTorch [FSDP](https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/) and [Deepspeed](https://github.com/microsoft/DeepSpeed).  This PR includes examples of FSDP and Deepspeed training jobs that utilize the SMDDP library. 

The directory structure of the `dataparallel` folder is now as follows:
* `fully_sharded_data_parallel` - contains Falcon example for FSDP training
* `deepspeed` - contains Llama example for Deepspeed training

Note that the Deepspeed example uses a dummy dataset generated within the training script itself and the FSDP example uses the pre-approved [SST2](https://huggingface.co/datasets/glue) dataset.  

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by sheilaliuxl — COMMENTED** [review]
Quickly went through the FSDP notebook, will do for Llama2 as well in another round.

Thank you!

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/requirements.txt
Comment (ruhanprasad) [reply training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/requirements.txt:5]: This example doesn't use parameter-efficient fine tuning (peft), but the notebook example this was built off used it in one of the files.  I will remove this from the `requirements.txt` as it is not needed

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/utils.py
Comment (ruhanprasad) [reply training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/utils.py:66]: thanks, removed this function as it is not called in `train.py`

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/utils.py
Comment (ruhanprasad) [reply training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/utils.py:52]: Yea it's weird, the FSDP documentation shows using `torch.save` as well here: https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html. 

For the optimizer state saving: Fair callout, I think since the focus here is more to show a simple model training example, I can add a comment that points to an example on how to save optimizer state as well.  The original example this script is based this off of saves the trained model weights only, which is desired if the model is only going to be used for inference afterwards and not additional training on the dataset.

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/SMDDP_FSDP_example.ipynb
Comment (ruhanprasad) [reply training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/SMDDP_FSDP_example.ipynb:256]: fixed, thanks

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py
Comment (ruhanprasad) [reply training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py:238]: adding additional condition for outer-loop as well

File: training/distributed_training/pytorch/data_parallel/deepspeed/llama2/code/utils.py
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/data_parallel/deepspeed/llama2/code/utils.py:42]: ditto to the FSDP example.

File: training/distributed_training/pytorch/data_parallel/deepspeed/llama2/code/train.py
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/data_parallel/deepspeed/llama2/code/train.py:162]: ditto

File: training/distributed_training/pytorch/data_parallel/deepspeed/llama2/SMDDP_Deepspeed_example.ipynb
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/data_parallel/deepspeed/llama2/SMDDP_Deepspeed_example.ipynb:14]: nit: We probably need a `.` at the end of the sentence.
- ditto the next line for `note 2`.

**Review by 0x6b64 — APPROVED** [review]
Looks like there might be some inconsistent formatting, which the style checker should catch. LGTM otherwise!

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py:260]: Do we need this check for `eval`? I think it can be removed.

File: training/distributed_training/pytorch/data_parallel/deepspeed/llama2/code/train.py
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/data_parallel/deepspeed/llama2/code/train.py:183]: ditto to FSDP.

**Review by sheilaliuxl — COMMENTED** [review]
- Why are we not moving the AR notebooks then?
- CI/CD is failing, probably we need to address them before getting approval from the repo owner.

Thanks for the changes, a few more minor questions.

File: training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py
Comment (ruhanprasad) [reply training/distributed_training/pytorch/data_parallel/fully_sharded_data_parallel/falcon/scripts/train.py:265]: If the `max_steps` argument is set, then we don't want the number of validation steps (i.e. the number of batches fed to model for validation) to exceed it as well.

File: training/distributed_training/pytorch/data_parallel/deepspeed/llama2/smddp_deepspeed_example.ipynb
Comment (sheilaliuxl) [inline training/distributed_training/pytorch/data_parallel/deepspeed/llama2/smddp_deepspeed_example.ipynb:104]: I find existing notebook adding examples of those placeholder vars, maybe we can do that as well.
- It's easier for the reader to follow then, e.g. (though I think the value is not really a valid example)

```
image = "<ADD NAME OF REPO>"  # Example: bert-smdataparallel-sagemaker (Change to a valid example/ format instead)
```

**Review by sheilaliuxl — APPROVED** [review]

**Review by atqy — APPROVED** [review]

**Pull request merged**
